### PR TITLE
Element/Extended Data Row additional visibilities

### PR DESCRIPTION
**Pull request by joeferner**
This is a big change so I marked the couple spots that I though were the most interesting.

- Accumulo stores the additional visibilties and the Accumulo iterator filters them out
- Elasticsearch also stores additional visibilties. ES then uses the metadata table to list all possible additional visibilities. When it goes to do a search it will iterate through the list to know which additional visibilties to include/exclude from the query.

### CODE REVIEW

File: accumulo/README.md [hunk 1]
```diff
@@ -18,6 +18,7 @@ The graph is stored in a hand full of tables with the following schema.
 <tr><td>V[id]</td>                                        <td>V</td>          <td>-</td>                                             <td>-</td>               <td>Vertex id</td></tr>
 <tr><td>V[id]</td>                                        <td>H</td>          <td>H</td>                                             <td>-</td>               <td>Vertex hidden marker</td></tr>
 <tr><td>V[id]</td>                                        <td>D</td>          <td>D</td>                                             <td>-</td>               <td>Vertex soft delete marker</td></tr>
+<tr><td>V[id]</td>                                        <td>AV</td>         <td>[visibility_string]</td>                           <td>-</td>               <td>Additional visibility to apply to the element</td></tr>
```
Comment (joeferner) [inline accumulo/README.md]: The additional visibilities are just stored here using the visibility string as the column qualifier

File: elasticsearch5/src/main/java/org/vertexium/elasticsearch5/Elasticsearch5SearchIndex.java [hunk 2]
```diff
@@ -95,6 +92,8 @@
     public static final String EXTENDED_DATA_TABLE_NAME_FIELD_NAME = "__extendedDataTableName";
     public static final String EXTENDED_DATA_TABLE_ROW_ID_FIELD_NAME = "__extendedDataRowId";
     public static final String EXTENDED_DATA_TABLE_COLUMN_VISIBILITIES_FIELD_NAME = "__extendedDataColumnVisibilities";
+    public static final String ADDITIONAL_VISIBILITY_FIELD_NAME = "__additionalVisibility";
```
Comment (joeferner) [inline elasticsearch5/src/main/java/org/vertexium/elasticsearch5/Elasticsearch5SearchIndex.java]: stored as multi-values in this field

File: elasticsearch5/src/main/java/org/vertexium/elasticsearch5/Elasticsearch5SearchIndex.java [hunk 3]
```diff
@@ -95,6 +92,8 @@
...
+    public static final String ADDITIONAL_VISIBILITY_METADATA_PREFIX = "elasticsearch_additionalVisibility_";
```
Comment (joeferner) [inline elasticsearch5/src/main/java/org/vertexium/elasticsearch5/Elasticsearch5SearchIndex.java]: stored in metadata table to know what to filter on in queries

File: core/src/main/java/org/vertexium/ExtendedDataRowBase.java [hunk 4]
```diff
@@ -7,6 +7,7 @@
 import org.vertexium.util.FilterIterable;
 
 import java.util.Iterator;
+import java.util.Set;
```
Comment (mwizeman) [inline core/src/main/java/org/vertexium/ExtendedDataRowBase.java]: Extra?

File: core/src/main/java/org/vertexium/Graph.java [hunk 5]
```diff
@@ -1221,7 +1235,31 @@ default void markEdgeVisible(Edge edge, Visibility visibility, Authorizations au
...
+        return getExtendedData(elementType, elementId, tableName, FetchHints.ALL, authorizations);
+    }
+
+    /**
+     * Gets the specified extended data rows.
+     *
+     * @param elementType    The type of element to get the rows from
+     * @param elementId      The element id to get the rows from
+     * @param tableName      The name of the table within the element to get the rows from
+     * @param authorizations The authorizations used to get the rows
```
Comment (mwizeman) [inline core/src/main/java/org/vertexium/Graph.java]: Missing the `fetchHints` param

File: core/src/main/java/org/vertexium/GraphBaseWithSearchIndex.java [hunk 6]
```diff
@@ -181,30 +182,28 @@ public SearchIndexSecurityGranularity getSearchIndexSecurityGranularity() {
...
                 elementsToAddToIndex.add(element);
-                //noinspection unchecked
-                extendedDataToIndex.add(new ElementAndIterableExtendedDataMutation(element, m.getExtendedData()));
             }
         }
         getSearchIndex().addElements(this, elementsToAddToIndex, authorizations);
-        for (ElementAndIterableExtendedDataMutation ed : extendedDataToIndex) {
-            getSearchIndex().addElementExtendedData(this, ed.element, ed.extendedData, authorizations);
+        for (ElementMutation<? extends Element> m : mutations) {
+            if (m.getIndexHint() == IndexHint.INDEX) {
```
Comment (mwizeman) [inline core/src/main/java/org/vertexium/GraphBaseWithSearchIndex.java]: If you flip the order of the `for` and the `if` you can save a little work by not having to iterate.

File: core/src/main/java/org/vertexium/mutation/ElementMutation.java [hunk 7]
```diff
@@ -152,6 +152,96 @@
...
+    default ElementMutation<T> addAdditionalVisibility(String visibility) {
+        return addAdditionalVisibility(visibility, null);
+    }
+
+    /**
+     * Adds additional visibilities to the element. These visibilities will not effect the Visibility returned from
+     * {@link Element#getVisibility()}. These visibilities can also be bypassed by {@link FetchHints}.
+     *
+     * @param visibility The additional visibility to be applied
+     */
```
Comment (mwizeman) [inline core/src/main/java/org/vertexium/mutation/ElementMutation.java]: Missing the `eventData` param

File: core/src/main/java/org/vertexium/mutation/ElementMutation.java [hunk 8]
```diff
@@ -152,6 +152,96 @@
...
+     * @param visibility The additional visibility to be deleted
+     */
+    default ElementMutation<T> deleteAdditionalVisibility(String visibility) {
+        return deleteAdditionalVisibility(visibility, null);
+    }
+
+    /**
+     * Deletes a previously set additional visibility.
+     *
+     * @param visibility The additional visibility to be deleted
```
Comment (mwizeman) [inline core/src/main/java/org/vertexium/mutation/ElementMutation.java]: Missing the `eventData` param

File: core/src/main/java/org/vertexium/mutation/ElementMutation.java [hunk 9]
```diff
@@ -152,6 +152,96 @@
...
+        return addExtendedDataAdditionalVisibility(tableName, row, additionalVisibility, null);
+    }
+
+    /**
+     * Adds additional visibilities to the row column. These visibilities will not effect the Visibility returned from
+     * {@link Element#getVisibility()}. These visibilities can also be bypassed by {@link FetchHints}.
+     *
+     * @param tableName            Name of the table the row resides in
+     * @param row                  The row
+     * @param additionalVisibility The additional visibility to be applied
```
Comment (mwizeman) [inline core/src/main/java/org/vertexium/mutation/ElementMutation.java]: Missing the `eventData` param

File: core/src/main/java/org/vertexium/mutation/ElementMutation.java [hunk 10]
```diff
@@ -152,6 +152,96 @@
...
+        String tableName,
+        String row,
+        String additionalVisibility,
+        Object eventData
+    );
+
+    /**
+     * Deletes a previously set additional visibility.
+     *
+     * @param additionalVisibility The additional visibility to be deleted
```
Comment (mwizeman) [inline core/src/main/java/org/vertexium/mutation/ElementMutation.java]: Missing params

File: core/src/main/java/org/vertexium/mutation/ElementMutation.java [hunk 11]
```diff
@@ -152,6 +152,96 @@
...
+        String row,
+        String additionalVisibility
+    ) {
+        return deleteExtendedDataAdditionalVisibility(tableName, row, additionalVisibility, null);
+    }
+
+    /**
+     * Deletes a previously set additional visibility.
+     *
+     * @param additionalVisibility The additional visibility to be deleted
```
Comment (mwizeman) [inline core/src/main/java/org/vertexium/mutation/ElementMutation.java]: Missing params

File: inmemory/src/main/java/org/vertexium/inmemory/InMemoryExtendedDataRow.java [hunk 12]
```diff
@@ -36,21 +43,41 @@ public ExtendedDataRowId getId() {
...
                 }
             }
+            row.additionalVisibilities.addAll(additionalVisibilities);
             return row;
         } finally {
             propertiesLock.readLock().unlock();
         }
     }
 
+    private boolean canReadAdditionVisibility(VisibilityEvaluator visibilityEvaluator) {
```
Comment (mwizeman) [inline inmemory/src/main/java/org/vertexium/inmemory/InMemoryExtendedDataRow.java]: rename to `canReadAdditionalVisibility`

File: core/src/main/java/org/vertexium/GraphBaseWithSearchIndex.java (hunk 6 above)
Comment (joeferner) [reply core/src/main/java/org/vertexium/GraphBaseWithSearchIndex.java]: Looks like the mutation itself has the fetch hints so can't switch this around

Comment (mwizeman) [reply core/src/main/java/org/vertexium/GraphBaseWithSearchIndex.java]: Yep. I missed that part.

**Comment by mwizeman** [comment]
Looks like the build is failing in Travis due to checkstyle.

**Pull request merged**